### PR TITLE
Features

### PR DESCRIPTION
**Pull request by LinuxMercedes**
- [x] Error handling/debug messages
- [x] SSL

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+use std::io::prelude::*;
+use std::io::BufReader;
+
+use std::default::Default;
+use irc::client::prelude::*;
+
+use std::thread::spawn;
+
+fn main() {
+    let channels = vec![format!("#hashpipe")];
```
Comment (euank) [inline src/main.rs]: nix `format!`, you probably wanted `.to_string()`

Comment (LinuxMercedes) [reply src/main.rs]: Cool...I should open a PR and update the example for `irc` since it also has a syntax error =/

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                            (@arg channels: -c --channels +takes_value "Channel(s) to speak in (if not in raw mode)")
+                            // NOTE: long() is a required workaround for parsing long options with
+                            // hyphens; see https://github.com/kbknapp/clap-rs/issues/321
+                            (@arg raw_out: -o long("--raw-out") "Echo everything from the IRC server directly")
+                            (@arg raw_in: -i long("--raw-in") "Interpret STDIN as raw IRC commands")
+                           ).get_matches();
+
+    let nick = matches.value_of("nick").unwrap_or("hashpipe").to_string();
+    let server = matches.value_of("server").unwrap().to_string();
+    let channels :Vec<String> = matches.value_of("channels").unwrap_or("#hashpipe").split(",").map(|x| x.to_string()).collect();
```
Comment (euank) [inline src/main.rs]: style: `: Vec`, not ` :Vec`

Comment (euank) [reply src/main.rs]: Also, you should install rustfmt and run it on all the rust you ever write

File: src/main.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    spawn (move || run_irc(irc_server, raw_out, sirc));
+
+    // Wait until we've joined all the channels we need to
+    let mut join_count = 0;
+    while join_count < channels.len() {
+        chan_select! {
+            signal.recv() -> _signal => {
+                server.send_quit("#|").unwrap();
+                return;
```
Comment (euank) [inline src/main.rs]: no way this return is needed, right?

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    Join,
+}
+
+fn main() {
+    // Catch signals we expect to exit cleanly from
+    let signal = chan_signal::notify(&[Signal::INT, Signal::TERM, Signal::PIPE]);
+
+    let matches = clap_app!
+        (hashpipe =>
+         (version: "0.1")
```
Comment (euank) [inline src/main.rs]: Cool trick @ianweller pointed out for version and author: https://github.com/euank/synapse-password-reset/blob/a7da24858cbcaf2af00fd3a8b82e1581e4d157f9/src/main.rs#L101-L102

**Comment by euank** [comment]
I'll toss on a LGTM apart from my version nit.

I'm sure there's other things we *could* do, but if it works it works and it looks pretty sane to me.

I do think we should think about integrating into the tokio ecosystem to make the async look a bit better, but that's definitely something for "later this decade" not "in this pr"

File: src/main.rs (hunk 4 above)
Comment (LinuxMercedes) [reply src/main.rs]: I was wondering if there was something like this!

**Comment by LinuxMercedes** [comment]
Cool. I've added my todo list to the PR (should have done that a while back). 

What do you think about the non-raw output? Currently it just displays PRIVMSGs; are there other things that would definitely be good to display?

**Comment by euank** [comment]
It might as well get notices and CTCP ACTION as well since those are both relatively common, but privmsgs should be the most important bit certainly.

**Comment by LinuxMercedes** [comment]
ACTION is techncially a PRIVMSG, so it "gets" those (although you have to write 0x01 bytes to send one, but that's a problem for another day). I've added NOTICE, though.

**Comment by LinuxMercedes** [comment]
@euank Does the README sufficiently document what users need to know?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# #|: Pipe data to and from IRC
+
+Hashpipe lets you connect stdin and stdout to IRC with Unix pipes. For example, `sleep 5; echo "done" | hashpipe -s irc.freenode.net` messages `#hashpipe` when a long-running command finishes.
```
Comment (euank) [inline README.md]: nit: "messages the `#hashpipe` irc channel on the freenode network". More explicit early on in the readme is good imo

**Review by euank — COMMENTED** [review]
just reviewed the readme. lgtm, though more examples would be cool!

For some cool examples, consider the following:

```bash
if hashpipe --server my.irc.server --channels "#commandcentre" | grep -m 1 "oh geez, it's the law"; then
  shred ~/Documents/secret/**/*
fi
```

Or

```bash
set -e
set -o pipefail
some-command-with-output | hashpipe -s my.irc.server --channels "#reports"
```

... others I'm sure

File: README.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+```
+USAGE:
+    hashpipe [FLAGS] [OPTIONS] --server <server>
+
+FLAGS:
+    -h, --help       Prints help information
+    -q               Only print errors (overrides -v; overridden by raw output)
+    -i, --raw-in     Interpret STDIN as raw IRC commands
+    -o, --raw-out    Echo everything from the IRC server directly
+    -e, --ssl        Enable SSL encryption
```
Comment (euank) [inline README.md]: let's default to ssl and have a flag to disable instead.

File: src/main.rs [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+    let raw_in = matches.is_present("raw_in");
+    let quiet = matches.is_present("quiet");
+
+    let nick = matches.value_of("nick").unwrap_or("hashpipe").to_string();
+    let server = matches.value_of("server").unwrap().to_string();
+    let ssl = matches.is_present("ssl");
+    let port = matches.value_of("port").and_then(|p| p.parse().ok());
+
+    let channels: Vec<String> = match matches.value_of("channels") {
+        Some(chans) => chans.split(",").map(|x| x.to_string()).collect(),
```
Comment (euank) [inline src/main.rs]: So, cool trick, I can set `--channels ""` and 'channels' here will end up being `[""]` not `[]`, so it'll think it needs to join a channel but totally can't.

Solution here is either validating channels or at the least stripping empty values.

Comment (LinuxMercedes) [reply src/main.rs]: ack, yeah, let me do that. Is it possible to have channels with prefixes other than #?

Comment (euank) [reply src/main.rs]: You didn't want to know this, but '#', '&', '+', '!',  https://github.com/ircdocs/modern-irc/blob/85e3b4161fe589e1fa541c24a93944ad18a30e97/info/rfc2811.txt#L143-L145

In practice, it's '#'

Comment (iliana) [reply src/main.rs]: (also don't forget channels with double hashes, such as ##linux)

Comment (LinuxMercedes) [reply src/main.rs]: Ok, so I should be able to validate chan names by checking that the first char of a name is one of those 4 characters?

I suppose I will also write code to handle the situation of a particular join failing, if I can.

File: README.md (hunk 6 above)
Comment (LinuxMercedes) [reply README.md]: Done!

**Review by euank — APPROVED** [review]

**Pull request merged**